### DISCUSSION TITLE
Allow integrations to set `has_entity_name` to False

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#1023] [Entity Models]
## Proposal

Our developer guidelines currently state new integrations *must set `has_entity_name` to `True`*. Also, name translations and automatic device class names do not work if `has_entity_name` is not `True`. The developer documentation also states the `has_entity_name` property will eventually be removed, and when that's done, `has_entity_name` is effectively unconditionally True. However, I suggest we adjust that to:
- The `has_entity_name` property is here to stay
- New integrations *must actively set* `has_entity_name` to either `True` or `False`
- Automatic device class names work if `has_entity_name` has been set by the integration - the point of this is to not accidentally change the name of old integrations' entities
- Name translations always work

## Background:

Architecture discussion<https://github.com/home-assistant/architecture/discussions/957>mentions two issues:

1. Some devices have the concept of "channels", where each channel may have a switch, various sensors and so on. To improve naming of for example sensors tied to a channel, [placeholders in entity names](https://github.com/home-assistant/core/pull/104453) help

2. In some cases it's not suitable to prefix entity names with the device name. Devices with channels is one example. A very different example is sensors for the `time_date` integration; if those sensors are collected under a device the resulting names will be "Time & Date Date", "Time & Date Date & Time" etc.

**Comment by gjohansson-ST** [comment]
Another option would perhaps be to use a new property for if device name is to be used but perhaps that is like opening pandoras box.
But it would allow for `has_entity_name` to eventually be removed (but obviously then we have another property/attribute).

**Comment by emontnemery** [comment]
This proposal was too focused on how to solve a problem (setting `has_entity_name` to False), and does not really focus on what the problem is and what we want to achieve. I'm closing this now and will open a new discussion.
